src/pybind/mgr/cephadm/ssh.py

```diff
@@ -135,16 +135,27 @@ async def _execute_command(self,
         cmd = "sudo " + " ".join(quote(x) for x in cmd)
         logger.debug(f'Running command: {cmd}')
         try:
-            r = await conn.run(cmd, input=stdin.decode() if stdin else None)
+            # The default value passed to encoding in asyncssh is an empty
+            # tuple. If an empty tuple is found when a channel is created, the
+            # encoding 'utf-8' is set automatically. To prevent that for
+            # `type(input) is bytes` we need to set it explicitly to None, so
+            # that no conversion whatsoever is done.
+            encoding = ()
+            if type(stdin) is bytes:
```

Comment:
per the type annotation, `stdin` can only be `bytes` or `None`. if that's the case, i'd prefer putting something like
```py
if stdin is None:
  encoding = `utf-8`
else:
  encoding = None
r = await conn.run(cmd, input=stdin, encoding=encoding)
```

for two reasons:

1. to be more explicit
2. `stdin` is None or a `bytes` object, so there is no need to put `stdin if stdin else None`.

Reply:
maybe it would be simpler to interact with asyncssh using [raw bytes](https://asyncssh.readthedocs.io/en/latest/api.html#asyncssh.SSHClientConnection.create_session)?

```
... `None` can be passed in if the application wishes to send and receive raw bytes
```

Reply:
after a second thought. yeah, that'd be a lot simpler!

Reply:
errr actually I don't think this will work, because some callers expect to pass input/output as str data (that will be utf-8 encode/decode) and yet others expect raw binary bytes data (see `_write_remote_file`).

transferring remote files (cephadm zip binary) via`asyncssh.scp` might make this implementation a bit cleaner: 
https://github.com/ceph/ceph/pull/43458